src/main/java/io/stargate/sgv2/jsonapi/api/model/command/clause/filter/FilterClause.java

```diff
@@ -12,6 +20,112 @@
+    String path = comparisonExpression.getPath();
+    boolean isPathIndexed = indexingProjector.isPathIncluded(path);
+
+    // If path is "_id" and it's denied, the operator can only be $eq or $in
+    if (path.equals(DocumentConstants.Fields.DOC_ID) && !isPathIndexed) {
+      FilterOperator filterOperator = comparisonExpression.getFilterOperations().get(0).operator();
+      // if operator is not $eq or $in, throw error
+      if (!filterOperator.equals(ValueComparisonOperator.EQ)
+          && !filterOperator.equals(ValueComparisonOperator.IN)) {
+        throw new JsonApiException(
+            ErrorCode.ID_NOT_INDEXED,
+            String.format(
+                "%s: The filter path ('%s') is not indexed, you can only use $eq or $in as the operator",
+                ErrorCode.ID_NOT_INDEXED.getMessage(), DocumentConstants.Fields.DOC_ID));
+      } else {
+        // else, _id can be used, return
+        return;
+      }
+    }
+
+    JsonLiteral<?> operand = comparisonExpression.getFilterOperations().get(0).operand();
+    // If path is an object (like address), validate the incremental path (like address.city)
+    if (operand.type() == JsonType.ARRAY || operand.type() == JsonType.SUB_DOC) {
+      if (operand.value() instanceof HashMap<?, ?> map) {
+        validateHashMap(indexingProjector, map, path);
+      }
+      if (operand.value() instanceof List<?> list) {
+        validateList(indexingProjector, list, path);
+      }
+    } else {
+      // If path is not an object and is not indexed, throw error
+      if (!isPathIndexed) {
+        throw new JsonApiException(
+            ErrorCode.UNINDEXED_FILTER_PATH,
+            String.format(
+                "%s: The filter path ('%s') is not indexed",
+                ErrorCode.UNINDEXED_FILTER_PATH.getMessage(), comparisonExpression.getPath()));
+      }
+    }
+  }
+
+  private void validateHashMap(
+      DocumentProjector indexingProjector, HashMap<?, ?> map, String currentPath) {
+    for (Map.Entry<?, ?> entry : map.entrySet()) {
+      if (entry.getKey() instanceof String) {
+        String subPath = (String) entry.getKey();
+        String incrementalPath = currentPath + "." + subPath;
+        if (!indexingProjector.isPathIncluded(incrementalPath)) {
+          throw new JsonApiException(
+              ErrorCode.UNINDEXED_FILTER_PATH,
+              String.format(
+                  "%s: The filter path ('%s') is not indexed",
+                  ErrorCode.UNINDEXED_FILTER_PATH.getMessage(), incrementalPath));
+        }
+        // continue build the incremental path if the value is a hashmap
+        if (entry.getValue() instanceof HashMap<?, ?> valueMap) {
+          validateHashMap(indexingProjector, valueMap, incrementalPath);
+        }
+        // continue build the incremental path if the value is a list
+        if (entry.getValue() instanceof List<?> list) {
+          validateList(indexingProjector, list, incrementalPath);
+        }
```

Comment:
Is the value not allowed to be a List? I used this test case: 
```
{
    "find": {
        "filter": {
            "address": {
                "street": {
                    "$in" :["xxx"]
                }
            }
        }
    }
}
```
and got the error message:
```
{
    "errors": [
        {
            "message": "Unsupported NodeType OBJECT",
            "errorCode": "UNSUPPORTED_FILTER_DATA_TYPE"
        }
    ]
}
```
The error message is not clear as well.

Reply:
Not good indeed.
I guess it goes into $date filter logic.
Let me raise a ticket.

Reply:
You can't use nesting like that; references need to be paths like `address.street`. And yes, exception is not very good, should figure out that `street` is not a valid Operator I think.

Reply:
Thank you!